### PR TITLE
Add TooltipBubble and supporting methods

### PR DESCRIPTION
**Pull request by somewhatabstract**
Here, we build on #146 and add the positioning component, `TooltipBubble`.

This component uses `react-popper` to provide the positioning support, adding our own visibility modifier built using the things we added in #145.

In a later update, this is likely to be split into a `TooltipBubblePositioner` component for the react-popper wrapping and a `TooltipBubble` component for styling within that positioning component.

This is part of breaking up #140.

### CODE REVIEW

**Comment by jeresig** [comment]
Deploy preview for *wonder-blocks* ready!

Built with commit 3e433a5cd944c440b6782409c28bf91e4e1df179

https://deploy-preview-148--wonder-blocks.netlify.com

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import {css, StyleSheet} from "aphrodite";
+import * as React from "react";
+import {Popper} from "react-popper";
+
+import visibilityModifierDefaultConfig from "../util/visibility-modifier.js";
+
+import type {PopperChildrenProps} from "react-popper";
+import type {Placement} from "../util/types.js";
+
+type Props = {|
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: Thanks for taking the plunge and using exact types.  I haven't been pressing people on it since the version of flow we're on has issues with spreading props.  Admittedly we don't do that that often.

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+export default class TooltipBubble extends React.Component<Props> {
+    _renderContent(childrenProps: PopperChildrenProps) {
+        const {children} = this.props;
+        const {
+            ref,
+            outOfBoundaries,
+            style,
+            placement,
+            //  arrowProps,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: Will `arrowProps` be making a comeback in a subsequent PR?

File: packages/wonder-blocks-tooltip/__mocks__/popper.js.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// @flow
+/**
+ *  Mock to allow us to run tests that need react-popper or popper.js
+ * Inspired by:
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/__mocks__/popper.js.js]: super nit: inconsistent spacing

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        } = childrenProps;
+
+        return (
+            <div
+                data-placement={placement}
+                ref={ref}
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: Can you add a comment as to why we're not using `View` here?

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        const resultRef = (((
+            <div ref={(ref) => this.props.resultRef(ref)}>
+                This is a pretend string with a ref so we can detect it being
+                rendered
+            </div>
+        ): any): string);
+        return (
+            <div>
+                <div ref={(ref) => this.updateRef(ref)}>Anchor</div>
+                <TooltipBubble
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/components/tooltip-bubble.test.js]: The tests should be written using `View` since people will be using `View` in production.

File: packages/wonder-blocks-tooltip/package.json [hunk 6]
```diff
@@ -15,6 +15,7 @@
   "author": "",
   "license": "MIT",
   "dependencies": {
+    "react-popper": "^1.0.0-beta.6",
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/package.json]: 🎵 living on the edge... 🎵

File: packages/wonder-blocks-tooltip/util/is-obscured.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    // going to end up hiding, so, you know, probably don't do that.
+    const anchorRect = element.getBoundingClientRect();
+    const style =
+        ((element: any).currentStyle: ?CSSStyleDeclaration) ||
+        window.getComputedStyle(element);
+    const anchorLeft =
+        anchorRect.left +
+        parseFloat(style.marginLeft) +
+        parseFloat(style.paddingLeft) +
+        parseFloat(style.borderLeftWidth);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.js]: Would this be affected by `box-sizing`?

File: packages/wonder-blocks-tooltip/util/is-obscured.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    // saying that we're visible if we hit a parent because we already
+    // checked them for visibility earlier.
+    const topLeftVisible =
+        topLeftElement &&
+        (element.contains(topLeftElement) || topLeftElement.contains(element));
+    const bottomRightVisible =
+        bottomRightElement &&
+        (element.contains(bottomRightElement) ||
+            bottomRightElement.contains(element));
+    return !topLeftVisible && !bottomRightVisible;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.js]: Make sense.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                </div>
+            );
+            mount(nodes);
+        };
+
+        const actAndAssert = (ref) => {
+            const element = ((ReactDOM.findDOMNode(ref): any): Element);
+            // When not obscurred, elementFromPoint should return the element.
+            // Flow doesn't like us doing this to the document $FlowFixMe
+            document.elementFromPoint = jest.fn().mockReturnValue(element);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Does this resent `elementFromPoint` after the test?  Not a big deal if it doesn't... I'm curious more than anything.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    test("element is obscured, returns true", (done) => {
+        // Arrange
+        let otherElement = null;
+        const arrange = (actAndAssert) => {
+            const nodes = (
+                <div>
+                    <div ref={(ref) => (otherElement = ref)}>
+                        Pretend this covers everything
+                    </div>
+                    <div ref={actAndAssert}>Unobscured</div>
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: nit: `Unobscured => Obscured`

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                </div>
+            );
+            mount(nodes);
+        };
+
+        const actAndAssert = (ref) => {
+            const element = ((ReactDOM.findDOMNode(ref): any): Element);
+            // When not obscurred, elementFromPoint should return the element.
+            // Flow doesn't like us doing this to the document $FlowFixMe
+            document.elementFromPoint = jest.fn().mockReturnValue(otherElement);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: We should have some tests where `elementFromPoint` returns `otherElement` and the unobscured element is a child of `otherElement` or vice versa.

File: packages/wonder-blocks-tooltip/util/visibility-modifier.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    // If we're hidden, we mimic what the built-in hide method does,
+    // and set the hide flag and the OOB attribute with appropriate
+    // short-circuiting.
+    if (hide) {
+        if (data.hide) {
+            return data;
+        }
+        data.hide = true;
+        data.attributes["x-out-of-boundaries"] = "";
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.js]: Is `"x-out-of-boundaries"` only used for test purposes?

File: packages/wonder-blocks-tooltip/util/visibility-modifier.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+    // First, we see how the element intersects with its scroll parents.
+    // If it doesn't, then we should hide it.
+    // Otherwise, we check to see if anything else obscures the component (like
+    // a fixed or absolute positioned element).
+    const {horizontal, vertical} = getElementIntersection(anchorElement);
+    const hide =
+        horizontal !== "within" ||
+        vertical !== "within" ||
+        isObscured(anchorElement);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.js]: Nice!

File: packages/wonder-blocks-tooltip/util/visibility-modifier.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        if (!data.hide) {
+            return data;
+        }
+        data.hide = false;
+        data.attributes["x-out-of-boundaries"] = false;
+    }
+
+    // Always have to return the data object to ensure the modifier chain
+    // in popper.js is unbroken.
+    return data;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.js]: Maybe link to the docs for popper.js modifiers in the docstring at the beginning fo this file.

File: packages/wonder-blocks-tooltip/util/visibility-modifier.test.js [hunk 15]
```diff
@@ -0,0 +1,185 @@
+// @flow
+import Popper from "popper.js";
+import * as Core from "@khanacademy/wonder-blocks-core";
+//import isObscured from "./is-obscured.js";
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.test.js]: remove

File: packages/wonder-blocks-tooltip/util/visibility-modifier.test.js [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        // that the results return what we want.
+        describe("Obscured by one or more scroll parents", () => {
+            test("Obscured horizontally only, is not visible", () => {
+                // Arrange
+                const psuedoAnchorElement = {
+                    getBoundingClientRect: () => ({
+                        top: 0,
+                        left: 0,
+                        right: 100,
+                        bottom: 100,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.test.js]: The value of this is unimportant since we're mocking `Core.getElementIntersection` anyways, right?

File: packages/wonder-blocks-tooltip/util/is-obscured.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    // Before we assume we're visible let's check to see if something else
+    // is obscuring us.  Here we make the assumption that if our topleft and
+    // bottomright corners are covered by something, we're not visible.
+    // There are ways that this can still not work, such as different
+    // elements only covering those points and the remainder being visible,
+    // or if some covering element has none for pointer-events style, but
+    // those edge cases shouldn't bother the main usages for this method.
+
+    // NOTE: If the anchor element has `pointer-events: none`, we're always
+    // going to end up hiding, so, you know, probably don't do that.
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.js]: I don't see any code that checks for `pointer-events: none`.  Could you explain more how this works?

File: packages/wonder-blocks-tooltip/util/visibility-modifier.test.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+                // Act
+                const result = visibilityModifierDefaultConfig.fn(data);
+
+                // Assert
+                expect(result.hide).toBeTruthy();
+                expect(result.attributes["x-out-of-boundaries"]).toBe("");
+            });
+        });
+
+        test("Obscured by fixed or absolute components, is not visible", () => {
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/visibility-modifier.test.js]: I see... we're making the assumption that by having `getElementIntersection` return `CompletelyWithin` and `isObscured` return `true` that we're simulating a similar situation to what would happen if the anchor was obscured by a fixed or absolute components.  Can you add a comment as it's not immediately obvious how the test title follows from the test?

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js (hunk 2 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: Yes, in fact, it's in PR #155

File: packages/wonder-blocks-tooltip/util/is-obscured.js (hunk 17 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.js]: `elementFromPoint` doesn't return things that have `pointer-events: none`. We could add a runtime check for that on our anchors but that seems a little heavy-handed for what amounts to a cornercase.

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js (hunk 1 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: It's habit now and that's worked out great. The cases where we don't want exact make themselves apparent really quickly.

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.test.js (hunk 5 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/components/tooltip-bubble.test.js]: Sadly, that creates a lot of heavy lifting with the refs since the popper.js stuff wants `HTMLElement` types.

In reality, this component will be internal to our other tooltip components, so I don't know that we gain anything by putting in the effort to use View here. Certainly, we should use public components for our public components (which for Tooltip, will be Tooltip and TooltipContent).

File: packages/wonder-blocks-tooltip/components/tooltip-bubble.js (hunk 4 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/components/tooltip-bubble.js]: Sadly, that creates a lot of heavy lifting with the refs since the popper.js stuff wants `HTMLElement` types. I've added a comment in the code.

File: packages/wonder-blocks-tooltip/package.json (hunk 6 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/package.json]: 😁

File: packages/wonder-blocks-tooltip/util/is-obscured.js (hunk 7 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.js]: I am not certain at this point. We should definitely test more edge cases (no pun intended).

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js (hunk 9 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Hmm, I think so, but I should check!

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: I've added some code to reset it, just in case.

File: packages/wonder-blocks-tooltip/util/visibility-modifier.js (hunk 12 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/visibility-modifier.js]: I've added a link in the code. Reproduced here:

https://github.com/FezVrasta/popper.js/blob/08c5d6010346bf9df06e9f81a54fa6c2c51e3639/packages/popper/src/modifiers/hide.js#L29-L42

File: packages/wonder-blocks-tooltip/util/visibility-modifier.js (hunk 14 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/visibility-modifier.js]: Good idea.

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/visibility-modifier.js]: https://popper.js.org/popper-documentation.html#modifiers

File: packages/wonder-blocks-tooltip/util/visibility-modifier.test.js (hunk 16 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/visibility-modifier.test.js]: Yeah, I can add a comment to make that clear.

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/visibility-modifier.test.js]: In fact, they're a hangover from before I split `isObscured` into it's own method. Now that we're mocking that method out, we don't need this fake bounds rect call. :grin:

File: packages/wonder-blocks-tooltip/util/is-obscured.js (hunk 17 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.js]: I have updated the commentary for this.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            document.elementFromPoint = jest.fn().mockReturnValue(element);
+
+            // Act
+            const result = isObscured(element);
+
+            // Assert
+            expect(result).toBeFalsy();
+            // Reset the document method to avoid side-effects.
+            // Flow doesn't like us doing this to the document $FlowFixMe
+            document.elementFromPoint = ogElementFromPoint;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: I'm a little shaky with how `beforeEach` and `afterEach` interact with async tests so I'm cool with handling the reset inside each `test`.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js (hunk 11 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: @somewhatabstract friendly ping.

**Comment by kevinbarabash** [comment]
@somewhatabstract friendly ping regarding the isObscured tests.  I had suggested in one of my comments of adding tests to check this situations where the unobscured element was a child of the other element or vice versa.  Thoughts?

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js (hunk 19 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Yes, I considered the before each and after each. I think they work fine for async (I think the after each only runs once done() or timeout occurs), but I wasn't sure a single line in each warranted adding them. I could go either way.

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: I've added them. I think it tidies things up considerably and appears to work.

**Comment by somewhatabstract** [comment]
@kevinbarabash I've added an extra test.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 20]
```diff
@@ -123,9 +132,80 @@ describe("isObscured", () => {
...
+            );
+            mount(nodes);
+        };
+
+        const actAndAssert = () => {
+            if (!elementRef || !parentRef || !childRef) {
+                // We're not ready to act or assert yet.
+                // We need all three refs to be in.
+                return;
+            }
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Wow.

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js [hunk 21]
```diff
@@ -123,9 +132,80 @@ describe("isObscured", () => {
...
+            expect(childElement).toBeTruthy();
+
+            // When not obscurred, elementFromPoint should return the element.
+            // So let's return the element for one corner but not the other.
             // Flow doesn't like us doing this to the document $FlowFixMe
-            document.elementFromPoint = ogElementFromPoint;
+            document.elementFromPoint = jest
+                .fn()
+                .mockImplementationOnce(() => parentElement)
+                .mockImplementationOnce(() => childElement);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Cool!

File: packages/wonder-blocks-tooltip/util/is-obscured.test.js (hunk 20 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: I wish Github would show more of the code.

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: I think if you drag select a group of lines, you can comment on the group and it should show them all.

Comment (somewhatabstract) [reply packages/wonder-blocks-tooltip/util/is-obscured.test.js]: Hmm, I am mistaken. :(
